aeppl/loglik.py

```diff
@@ -87,30 +91,68 @@ def loglik(
 
     rng, size, dtype, *dist_params = rv_node.inputs
 
-    # Here, we plug the actual random variable into the log-likelihood graph,
-    # because we want a log-likelihood graph that only contains
-    # random variables.  This is important, because a random variable's
-    # parameters can contain random variables themselves.
-    # Ultimately, with a graph containing only random variables and
-    # "deterministics", we can simply replace all the random variables with
-    # their value variables and be done.
-
-    # tmp_rv_values = rv_values.copy()
-    # tmp_rv_values[rv_var] = rv_var
-
-    logpdf_var = _logpdf(rv_node.op, rv_value_var, *dist_params, **kwargs)
-
     # Replace random variables with their value variables
     replacements = rv_values.copy()
     replacements.update({rv_var: rv_value, rv_value_var: rv_value})
 
-    (logpdf_var,), _ = rvs_to_value_vars(
-        (logpdf_var,),
-        initial_replacements=replacements,
-    )
+    from collections import deque
+
+    # TODO: Walk the graph from its inputs to its outputs and construct the
+    # log
+    # TODO: Accept more than one graph.
+    # logp_fn = FunctionGraph(outputs=[rv_var], clone=False)
+
+    q = deque(io_toposort(graph_inputs((rv_var,)), (rv_var,)))
+
+    logpdf_var = None
+
+    while q:
+        node = q.popleft()
+
+        if not any(o in rv_values for o in node.outputs) and rv_var not in node.outputs:
+            if isinstance(node.op, RandomVariable):
+                warnings.warn(
+                    "Found a random variable that was neither among the observations "
+                    f"nor the conditioned variables: {node}"
+                )
+            continue
+
+        if isinstance(node.op, RandomVariable):
+            q_rv_var = node.outputs[1]
+            q_rv_value_var = replacements[q_rv_var]
+
+            # Replace `RandomVariable`s in the inputs with value variables.
+            # Also, store the results in the `replacements` map so that we
+            # don't need to redo these replacements.
+            value_var_inputs, _ = rvs_to_value_vars(
+                node.inputs,
+                initial_replacements=replacements,
+            )
+
+            q_logpdf_var = _logpdf(node.op, q_rv_value_var, *value_var_inputs, **kwargs)
+
+        else:
+            # Let's try deriving a log-likelihood from this
+            # non-`RandonVariable` node
+            q_logpdf_parts = []
+            for q_rv_var in node.outputs:
```

Comment:
I am not very familiar with multi-output nodes, is it safe to say the loglike can always be dealt independently for these?

Reply:
This loop should handle all cases just fine.  The important thing is that we make sure to only add the log-likelihoods for the outputs that are supposed to be in the log-likelihood we're computing.  The `replacements` check is supposed to do that, but, since other things could be in there, it might not be the best approach.

I'm about to push a docstring update to `loglik` that should explain exactly what this function is supposed to do (and why we should probably rename it).